### PR TITLE
fqdn: Pre-allocate ToFQDN selector identities

### PR DESCRIPTION
**Pull request by gandro**
This PR contains a further performance improvement for https://github.com/cilium/cilium/pull/32769, namely pre-allocation for selector identities as proposed in [CFP-28427](https://github.com/cilium/design-cfps/blob/fe6dd4faf9c03626428e63536d0f5c5ec6e82ef1/cilium/CFP-28427-fqdn-selectors.md#fqdn-identity-pre-allocation).

The basic idea is that we pre-allocate a matching identity whenever a new ToFQDN selector is added by the policy engine. This reduces critical path latency for DNS lookups, as no policy maps need to be updated when the first DNS lookup occurs.

See commit messages for details

### CODE REVIEW

File: pkg/fqdn/identities.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+func (a *identityAllocator) allocate(selector api.FQDNSelector) error {
+	id, ok := a.preAllocatedIdentities[selector]
+	if ok {
+		return errors.New("attempted to pre-allocate already pre-allocated identity")
+	}
+
+	lbls := labels.FromSlice([]labels.Label{selector.IdentityLabel()})
+	id, _, err := a.backend.AllocateLocalIdentity(lbls, true, identity.IdentityUnknown)
```
Comment (squeed) [inline pkg/fqdn/identities.go]: How bad would it be to pass `notifyOwners: false`, then batch all updates to the SelectorCache ourselves?

Comment (gandro) [reply pkg/fqdn/identities.go]: I'll look into it

Comment (squeed) [reply pkg/fqdn/identities.go]: On second thought, let's not do this. We might wind up in a bad race condition where IPs get an identity that is not yet plumbed in.

We should keep `notifyOwner: true`

**Comment by gandro** [comment]
So the approach of using `notifyOwner: true` does not work well, as we're updating the policy maps for each ToFQDN selector separately rather than batching. And even without batching we actually suffer from the race above (https://github.com/cilium/cilium/pull/33068#discussion_r1636022478)

I've been playing around with an alternative implementation where we allocate identities directly in the selector cache (and have the selector cache update itself accordingly), but that doesn't work, due to the fact that all FQDN identities are also selected by any rule selecting `world`.

Allocating during a FQDN selector update and relying on the caller(s) to recompute policy maps and eventually regenerate the endpoints for which the policy was updated works for those endpoints - but not for the endpoints which select the new identities indirectly via `world`. So basically if we want to be correct, we have to update the policy maps for _all endpoints_ whenever we register a new selector.

I don't want to trigger a "regenerate all endpoints" event from the SelectorCache because it seems like a very leaky abstraction making the code even more complicated. Thus, I'm currently tending on keeping the approach of doing things in the NameManager.

Doing it asynchronously from the NameManager however means we ought take care of two complications:

1. We need to block DNS lookups for that selector until the policy maps have been updated (to avoid the race you discovered above). But we probably need to this in any case, even if part of the logic lives were to live in selector cache - as soon as the FQDN selector is registered with the name manager, we need to expect lookups for it to happen, which can happen before we even finished distilling the policy.

2. Because identity allocation happens in the NameManager asynchronously, we probably update the policy map for some endpoints up to three times: First when the regenerate the endpoint due to one of its policies changing, then again once we've allocated the FQDN identity, and then a third time if the policy also contained CIDR rules that caused new CIDR identities to be allocated. But I really don't see how we can consolidate this

**Comment by gandro** [comment]
Closing this as not planned for now. This could be done, but it is an impactful change that is probably not worth it at the moment.

**Pull request closed without merging**